packages/react-components/src/components/timegraph-output-component.tsx

```diff
@@ -384,7 +500,105 @@ export class TimegraphOutputComponent extends AbstractTreeOutputComponent<Timegr
+        const filteredEvents = rangeEvents.filter(event => this.annotationMarkers?.includes(event.category));
+
+        filteredEvents.forEach(event => {
+            const categoryName = event.category;
+            if (!markers.has(categoryName)) {
+                markers.set(categoryName, []);
+            }
+            const states = markers.get(categoryName) || [];
+            const state = {
+                id: '1',
+                range: {
+                    start: event.range.start,
+                    end: event.range.end
+                },
+                label: event.label,
+                data: {
+                    style: event.data?.style,
+                    annotation: event
+                }
+            };
+            states.push(state);
+        });
+
+        const defaultRow = {
+            id: 1,
+            name: '',
+            range: {
+                start: this.props.viewRange.getstart(),
+                end: this.props.viewRange.getstart() + this.props.unitController.absoluteRange
+            },
+            states: [],
+            annotations: [],
+            prevPossibleState: Number.MIN_VALUE,
+            nextPossibleState: Number.MAX_VALUE
+        };
+        rows.push(defaultRow);
+
+        annotationEntries.push({
+            id: 0,
+            labels: [''],
+            parentId: -1
+        });
+
+        Array.from(markers.entries()).forEach((value, index) => {
+            annotationEntries.push({
+                id: index + 1,
+                labels: [value[0]],
+                parentId: 0
+            });
+
+            const row = {
+                id: 1,
+                name: '',
+                range: {
+                    start: this.props.viewRange.getstart(),
+                    end: this.props.viewRange.getstart() + this.props.unitController.absoluteRange
+                },
+                states: value[1],
+                annotations: [],
+                prevPossibleState: Number.MIN_VALUE,
+                nextPossibleState: Number.MAX_VALUE
+            };
+            rows.push(row);
+        });
+
+        this.markersData = {
+            rows,
+            range: newRange,
+            resolution: newResolution
+        };
+
+        this.setState({ annotationEntries });
+    }
+
+    private async fetchMarkersData(range: TimelineChart.TimeGraphRange, resolution: number) {
+        return {
+            rows: this.markersData ? this.markersData.rows : [],
+            range: this.markersData ? this.markersData.range : range,
+            resolution: this.markersData ? this.markersData.resolution : resolution
+        };
+    }
+
     private getStateStyle(state: TimelineChart.TimeGraphState) {
+
+        if (state.data && state.data.annotation) {
+            const style = this.getAnnotationStyle(state.data.annotation);
+            return {
+                color: style ? style.color : 0x000000,
+                opacity: style ? style.opacity : 1.0,
```

Comment:
In Trace Compass we don't use the opacity in the marker axis (only in the time graph itself), so they are fully opaque. If we want to achieve the same, you can just omit the opacity.

Reply:
This might actually be required because for marker sets, sometimes alternate markers (e.g. odd markers) have the same common marker color as the even marker, but with 0.0 opacity. That would make them invisible in the marker row. To be checked...